### PR TITLE
Reducing array allocations in expression interpreter

### PR DESCRIPTION
**Pull request by bartdesmet**
This addresses the low hanging fruit described in #10520. See the issue for a report on performance improvements.

### CODE REVIEW

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.Generated.cs [hunk 1]
```diff
@@ -18,52 +18,6 @@ internal partial class CallInstruction
         private const int MaxHelpers = 3;
         private const int MaxArgs = 3;
 
-        public virtual object InvokeInstance(object instance, params object[] args)
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.Generated.cs]: Note that these were dead code.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.Generated.cs [hunk 2]
```diff
@@ -250,12 +204,6 @@ public ActionCallInstruction(MethodInfo target)
             _target = (Action)target.CreateDelegate(typeof(Action), target);
         }
 
-        public override object Invoke()
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.Generated.cs]: Note that all logic for these is in the `Run` method; the `Invoke` method is never called . Also, this stuff is conditionally compiled using `FEATURE_FAST_CREATE` which is not used at all today.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.cs [hunk 3]
```diff
@@ -324,133 +324,91 @@ internal MethodInfoCallInstruction(MethodInfo target, int argumentCount)
...
-        public override object Invoke()
-        {
-            return InvokeWorker();
-        }
-        public override object Invoke(object arg0)
-        {
-            return InvokeWorker(arg0);
-        }
-        public override object Invoke(object arg0, object arg1)
+        public override int Run(InterpretedFrame frame)
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/CallInstruction.cs]: Removed indirections through virtual `Invoke` methods here. If we want to support fast paths for N-ary calls using delegates at some point - provided Core supports this - we can emit the specialized `ActionCall` and `FuncCall` instructions, which perform direct optimized stack manipulation, instead.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs [hunk 4]
```diff
@@ -2400,11 +2400,11 @@ private void CompileNewExpression(Expression expr)
 
                 if (updaters != null)
                 {
-                    _instructions.EmitByRefNew(node.Constructor, updaters.ToArray());
+                    _instructions.EmitByRefNew(node.Constructor, parameters, updaters.ToArray());
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/LightCompiler.cs]: Analogous to how we call `EmitCall` in the compiler, avoiding a second call to `GetParameters`.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs [hunk 5]
```diff
@@ -49,29 +49,27 @@ public override int Run(InterpretedFrame frame)
...
         public override string InstructionName
         {
             get { return "New"; }
         }
         public override int Run(InterpretedFrame frame)
         {
-            object[] args = new object[_argCount];
-            for (int i = _argCount - 1; i >= 0; i--)
-            {
-                args[i] = frame.Pop();
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/TypeOperations.cs]: Made analogous to `CallInstruction`; there's no need to do a bunch of `Pop`s which decrement the `StackIndex` if we can just obtain the arguments from the frame. (Note that `Pop` doesn't clear frame slots; it just reads and adjusts the pointer).

**Comment by bartdesmet** [comment]
CC @VSadov

**Comment by VSadov** [comment]
Nice! 
LGTM

**Pull request merged**